bigtable/client/detail/bulk_mutator_test.cc

```diff
@@ -0,0 +1,379 @@
+          }));
+
+  auto policy = bt::DefaultIdempotentMutationPolicy();
+  bt::detail::BulkMutator mutator("foo/bar/baz/table", *policy, std::move(mut));
+
+  // This test is simulating the expected behavior from the BulkApply() member
+  // function.
+  for (int i = 0; i != 2; ++i) {
+    EXPECT_TRUE(mutator.HasPendingMutations());
+    grpc::ClientContext context;
+    auto status = mutator.MakeOneRequest(stub, context);
+    EXPECT_TRUE(status.ok());
+  }
+  auto failures = mutator.ExtractFinalFailures();
+  ASSERT_EQ(1UL, failures.size());
+  EXPECT_EQ(1, failures[0].original_index());
+  EXPECT_EQ("bar", failures[0].mutation().row_key());
+  EXPECT_EQ(grpc::StatusCode::OUT_OF_RANGE, failures[0].status().error_code());
+}
+
+/// @test Verify that MultipleRowsMutator handles a stream with partial results
+TEST(MultipleRowsMutatorTest, PartialStream) {
+  namespace btproto = ::google::bigtable::v2;
+  namespace bt = bigtable;
+  using namespace ::testing;
+
+  // We are going to test the case where the stream does not contain a response
+  // for all requests.  Create a BulkMutation with two entries.
+  bt::BulkMutation mut(
+      bt::SingleRowMutation("foo", {bigtable::SetCell("fam", "col", 0, "baz")}),
+      bt::SingleRowMutation("bar",
+                            {bigtable::SetCell("fam", "col", 0, "qux")}));
+
+  // This will be the stream returned by the first request.  It is missing
+  // information about one of the mutations.
+  auto r1 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r1, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse* r) {
+        auto& e0 = *r->add_entries();
+        e0.set_index(0);
+        e0.mutable_status()->set_code(grpc::OK);
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r1, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  // The BulkMutation should issue a second request, this is the stream returned
+  // by the second request, which indicates success for the missed mutation
+  // on r1.
+  auto r2 = absl::make_unique<MockReader>();
+  EXPECT_CALL(*r2, Read(_))
+      .WillOnce(Invoke([](btproto::MutateRowsResponse* r) {
+        {
+          auto& e = *r->add_entries();
+          e.set_index(0);
+          e.mutable_status()->set_code(grpc::OK);
+        }
+        return true;
+      }))
+      .WillOnce(Return(false));
+  EXPECT_CALL(*r2, Finish()).WillOnce(Return(grpc::Status::OK));
+
+  btproto::MockBigtableStub stub;
+  EXPECT_CALL(stub, MutateRowsRaw(_, _))
+      .WillOnce(Invoke(
+          [&r1](grpc::ClientContext*, btproto::MutateRowsRequest const&) {
+            return r1.release();
+          }))
+      .WillOnce(Invoke(
+          [&r2](grpc::ClientContext*, btproto::MutateRowsRequest const&) {
+            return r2.release();
+          }));
+
+  auto policy = bt::DefaultIdempotentMutationPolicy();
+  bt::detail::BulkMutator mutator("foo/bar/baz/table", *policy, std::move(mut));
+
+  for (int i = 0; i != 2; ++i) {
```

Comment:
Add a comment here that 2 is because of `reader1` and `reader2`?

Reply:
Done.